cs/src/Connections/TunnelHost.cs

```diff
@@ -13,220 +13,183 @@
-                    }
-                }
-            });
+        // Capture the remote forwarder for the session id / remote port pair.
+        // This is needed later to stop forwarding for this port when the remote forwarder is disposed.
+        // Note when the client tries to open an SSH channel to PFS, the port forwarding service checks
+        // its remoteConnectors whether the port is being forwarded.
+        // Disposing of the RemotePortForwarder stops the forwarding and removes the remote connector
+        // from PFS.remoteConnectors.
+        RemoteForwarders.TryAdd(
+            new SessionPortKey(sessionId, (ushort)forwarder.RemotePort),
+            forwarder);
+        return;
+    }
 
-            return portDeleted;
-        }
 
-        /// <inheritdoc />
-        public async Task<TunnelPort> UpdatePortAsync(
-            TunnelPort updatedPort,
-            CancellationToken cancellation)
+    /// <inheritdoc />
+    public async Task RefreshPortsAsync(CancellationToken cancellation)
+    {
+        if (Tunnel == null || ManagementClient == null)
         {
-            if (Tunnel == null || Tunnel.Ports == null)
-            {
-                throw new InvalidOperationException("Tunnel must be running and have ports to update");
-            }
-
-            var port = await ManagementClient!.UpdateTunnelPortAsync(Tunnel, updatedPort, null, cancellation);
-            return port;
+            return;
         }
 
-        internal async Task ForwardPortAsync(
-            PortForwardingService pfs,
-            TunnelPort port,
-            CancellationToken cancellation)
-        {
-            var sessionId = Requires.NotNull(pfs.Session.SessionId!, nameof(pfs.Session.SessionId));
+        var updatedTunnel = await ManagementClient.GetTunnelAsync(
+            Tunnel, new TunnelRequestOptions { IncludePorts = true });
 
-            var portNumber = (int)port.PortNumber;
+        var updatedPorts = updatedTunnel?.Ports ?? Array.Empty<TunnelPort>();
+        Tunnel.Ports = updatedPorts;
 
-            if (pfs.LocalForwardedPorts.Any((p) => p.LocalPort == portNumber))
+        foreach (var port in updatedPorts)
+        {
+            foreach (var session in SshSessions
+                .Where((s) => s.IsConnected && s.SessionId != null))
             {
-                // The port is already forwarded. This may happen if we try to add the same port twice after reconnection.
-                return;
+                var key = new SessionPortKey(session.SessionId!, port.PortNumber);
+                if (!RemoteForwarders.ContainsKey(key))
+                {
+                    // Overlapping refresh operations could cause duplicate forward requests to be
+                    // sent to clients, but clients should ignore the duplicate requests.
+                    var pfs = session.GetService<PortForwardingService>() !;
+                    await ForwardPortAsync(pfs, port, cancellation);
```

Comment:
Calls to `ForwardPortAsync()` should be done concurrently. I'll update this code to add tasks to a list and then wait on all of them at once.

Reply:
Done.